packages/e2e-tests/specs/editor/various/links.test.js

```diff
@@ -335,16 +333,21 @@ describe( 'Links', () => {
 		// Deselect the link text by moving the caret to the end of the line
 		// and the link popover should not be displayed.
 		await page.keyboard.press( 'End' );
-		expect( await page.$( '.block-editor-url-popover' ) ).toBeNull();
+		expect( await page.$( '.components-popover__content .block-editor-link-control' ) ).toBeNull();
 
 		// Move the caret back into the link text and the link popover
 		// should be displayed.
 		await page.keyboard.press( 'ArrowLeft' );
-		expect( await page.$( '.block-editor-url-popover' ) ).not.toBeNull();
+		expect( await page.$( '.components-popover__content .block-editor-link-control' ) ).not.toBeNull();
 
 		// Press Cmd+K to edit the link and the url-input should become
 		// focused with the value previously inserted.
 		await pressKeyWithModifier( 'primary', 'K' );
+		await page.waitForSelector( '.components-popover__content .block-editor-link-control' );
+		// Click on the Edit button
+		const [ editButton ] = await page.$x( '//button[text()="Edit"]' );
+		await editButton.click();
```

Comment:
Do we want to make this change in behavior?

Notably, before these changes, upon pressing <kbd>Cmd+K</kbd>, the user would be able to edit the link immediately. Now, Cmd+K basically acts as a way to shift focus into the Popover attached to the link in the paragraph, but still in the "view"/"read-only" mode.

Before|After
---|---
![cmd-k-before](https://user-images.githubusercontent.com/1779930/73472476-1a437400-4359-11ea-8c9c-d6248b815af4.gif)|![cmd-k-after](https://user-images.githubusercontent.com/1779930/73472475-1a437400-4359-11ea-963d-87417d8ae0b6.gif)

It might be an okay change to make, if at least a conscious one.

I see both pros and cons:

- Con: More steps to get to the point of editing a link by keyboard alone.
- Pro: Avoids needing to create a way to forcefully control LinkControl's internal state (something like a `forceIsEditing` prop), which itself could lead to more consistency in using this component in general.
- Pro: If one _wanted_ to get to the "view" mode details of a link by keyboard alone, this change would make that workflow easier.

Reply:
I wasn't aware of this previously, but apparently <kbd>Ctrl+K</kbd> is pretty well-established as being associated with the intention of editing a link. So we probably want to retain the previous behavior.

>control-K is often used to add, edit, or modify a hyperlink to a Web page

https://en.wikipedia.org/wiki/Control-K

The article notes Microsoft Word, and I've also confirmed it works this way in Google Docs.

Reply:
> I wasn't aware of this previously, but apparently Ctrl+K is pretty well-established as being associated with the intention of editing a link. So we probably want to retain the previous behavior.

There's a bit of a conflict here that we need to sort out.

Consider the following:

1. Insert a paragraph with some text
2. Select some or all of that text
3. Press <kbd>Cmd+K</kbd> to start applying a link
4. Input a URL and press <kbd>Enter</kbd>

Where should focus be at this point?

In the current plugin, the focus is moved back to the paragraph, so you can proceed to continue editing the text of the paragraph. I've observed this behavior in Google Docs as well.

But there's a problem with this for the new link editing experience, which is: How do I toggle "Open in New Tab" if I'm only using the keyboard? The only way for me to shift focus into the link popover is using <kbd>Cmd+K</kbd>, but this immediately puts me into editing the URL, and there's no way for me to access the "Open in New Tab" toggle, because we (no longer) show the toggle when editing the URL. 

The options I might consider are:

- <kbd>Cmd+K</kbd> should just shift focus into the Popover. This breaks some widespread conventions / expectations around how <kbd>Cmd+K</kbd> tends to work, based on my [earlier comment](https://github.com/WordPress/gutenberg/pull/19462#discussion_r373156209).
- Or: Pressing <kbd>Enter</kbd> when editing a link should keep focus within the Popover. This could work, but it could slow the workflow for those who might expect to be able to proceed with their writing immediately upon pressing <kbd>Enter</kbd> (I would expect pressing <kbd>Escape</kbd> at this point would be enough to put the caret back into the paragraph). There's also a technical challenge here, in that the formatting implementation seems to become confused when focus stays within the popover after the new link is applied.
- Or: <kbd>Cmd+K</kbd> can continue to immediately start editing the link, and <kbd>Enter</kbd> once finished editing a link will still shift focus back to the paragraph, but then we also add an additional way to be able to navigate into the Popover from an active link without triggering it as an edit.

I was curious how Google Docs handles this, since they have a similar "popup", their behavior of <kbd>Cmd+K</kbd> is to immediately edit the URL, and their behavior of <kbd>Enter</kbd> once finished editing a URL is to shift focus back into the paragraph:

![g-docs-cmd-k](https://user-images.githubusercontent.com/1779930/73575368-09bcf780-4446-11ea-9024-7e914f8c8cd8.gif)

I'm actually not sure how one would be expected to access those "Copy Link" and "Remove Link" buttons in the popup using only keyboard:

![image](https://user-images.githubusercontent.com/1779930/73575439-2fe29780-4446-11ea-876c-0eb6838e1afc.png)

There's a Keyboard Shortcuts panel accessible pressing <kbd>Cmd+?</kbd>, and it includes an entry for "Move focus to popup", but the key combination does not appear to work for these link popups.

![image](https://user-images.githubusercontent.com/1779930/73575493-51dc1a00-4446-11ea-843f-a59c9e0f00e4.png)

Maybe it's assumed since there are other means to Remove or Copy a link in Google Docs, it's not necessary to have these buttons be accessible via keyboard. In our case, we're not able to make the same compromise, because there's no other way to toggle "Opens in New Tab".

Reply:
I don't feel like I'm capable of choosing which behavior makes the most sense. 

My cartesian logic (and technical one probably) prefers that this popover behaves like any other popover which is:

 - Ctrl + k or click to open popover in its initial state (input for new links, view for edited links)
 - Enter in the input mode keeps focus in the popover
 - Escape moves focus back to the paragraph

I believe this behavior is also good a11y wise but I understand that this could be confusing.

Any thoughts here @jasmussen @karmatosed

Reply:
Excellent conversation here. I would love input from skilled keyboard users on what is the most intuitive and elegant here. @MarcoZehe if you have time, perhaps? Your input would be valuable.

Google Docs and the popover that appears when you select a link does not seem like the best pattern to emulate as, and Andrew touches on this, focus remains in text, which means that pressing "tab" inserts a tab in the text.

However both the classic editor, and the specific ⌘K dialog in Google Docs, both seem like they could provide a pattern to follow here as well. In my reading, if focus is in an input field, pressing Enter closes the dialog and brings focus back to text. If focus is on a button, it toggles that button. With that heuristic, the following could work:

- Select text, press ⌘K, paste URL, press Enter and you're back in text.
- Select text, press ⌘K, paster URL, press Tab to set focus on the advanced dropdown. Press enter to open it. Press tab to set focus on "Open in new window". Press shift tab to go back to the Submit button, press Enter and you're back in text. 

To my understanding, this is also the behavior of the classic editor:

![classic](https://user-images.githubusercontent.com/1204802/73640733-1a9c8180-466f-11ea-8445-4b77cfe27d95.gif)

It's entirely possible I'm missing some context from the new link dialog, but if not, does the above heuristic make sense?

Reply:
> Select text, press ⌘K, paster URL, press Tab to set focus on the advanced dropdown. Press enter to open it. Press tab to set focus on "Open in new window". Press shift tab to go back to the Submit button, press Enter and you're back in text.

The problem is that this behavior won't work here do the design of the new LinkControl component. The component is now a "two steps" popover, you enter the link, once you submit it, you choose other options (new tab...).

This new component can be used today in the navigation and buttons block.

Reply:
I see it now. Took it for a spin:

![new behavior](https://user-images.githubusercontent.com/1204802/73642764-92b87680-4672-11ea-97d8-87415f7e1b75.gif)

To me this suggests the primary challenge is: how do you enter the popover that appears when the link has been created? Strawman suggestion: why doesn't Tab enter that dialog? Since it doesn't add tabs in the text, it might as well tab into that dialog.

Reply:
Keyboard focus can only be in one place at a time, and tab should always act predictably. Besides, why that extra step if I don't even want to set extra options? That sounds to me like an anti-pattern. So, I want to select text, press CTRL+K or CMD+K, paste the URL, enter to submit, or tab to the submit button and hit Enter or Space, and be back in the text to continue working. Shoving another dialog in my face to set options I might not even need would annoy me big times as a user, and also confuse me. That previous Link dialog that's also in the Classic editor where I can expand the advanced options if I need to, worked fine. If it must be a separate popover, make it a button in the main popover that opens this secondary one, and then submits the link with these set options once I okay/submit that secondary popover. Otherwise, keep the interaction as simple and straight-forward as possible.

Reply:
I admit I totally missed the replies here before merging (GitHub threads can be hard to follow, especially for such a large pull request). Trying to follow the conversation as it's evolved here, I want to make sure that nothing is lost for consideration and that we don't regress behaviors. I'm also conscious that this link experience is being proposed to support more advanced use cases of the Navigation Link and Buttons blocks, and brought to the paragraph to ensure a consistent experience for editing links.

From what I'm gathering in this discussion, one thing that seems important is to restore the workflow that <kbd>Cmd+k</kbd> immediately puts the user into editing a link, and ideally to treat "Opens in New Tab" as part of a set of advanced options. I think it might require something of a rethink for how we present the link preview when the current selection is within a link. With the discussion here, it feels that it might be problematic to include the "Open in New Tab" as part of this preview, when it's more of an advanced editing feature.

I will make sure to follow up with relevant tasks and/or pull requests.

Reply:
To clarify how the current behavior will be after merging this pull request:

1. Select some text
2. Press <kbd>Cmd+K</kbd>
3. Now focus is inside the popover in its preview state. From here, the user can tab to the "Edit" button or to the "Open in New Tab" button
4. After clicking "Edit", the user can change the URL
5. Pressing <kbd>Enter</kbd> to submit the changes will place focus back in the paragraph

To me, it seems the most problematic of these is the 3rd step, where pressing <kbd>Cmd+K</kbd> will shift focus into the preview popover, but there's still another step required before the user is able to edit the link. This was necessary, however, since in the current implementation of this link component, it's the only place where "Open in New Tab" can be toggled, and otherwise the option would not be reachable by keyboard. The alternative where focus remains in the popup after pressing <kbd>Enter</kbd> seems that it would be equally problematic. 

It appears the problem may be more about where the "Open in New Tab" toggle is placed. If it's something we think should be part of the link editing workflow (optionally as an advanced step), then I don't think it makes sense that it's shown as part of this preview state. Pressing <kbd>Cmd+K</kbd> should allow me to immediately edit a link, and I should still be able to toggle the option to "Open in New Tab" without also breaking my expectation that submitting my changes by pressing <kbd>Enter</kbd> would put the focus back in the paragraph.

This is my understanding of the problem, but I would appreciate any thoughts or corrections. I would also require some design help to accommodate this change.

Reply:
Thanks for all your careful consideration here, @aduth. I agree with @MarcoZehe about keeping the flow as simple as possible while allowing for access to the advanced options if necessary.

Reply:
To make sure this is tracked, I created an issue at #20007 to follow up on work related to this discussion.